## Peer-graded Assignment: 
## Segmenting and Clustering Neighborhoods in Toronto
### Importing Libraries 

In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto based on the postalcode and borough information.. However, unlike New York, the neighborhood data is not readily available on the internet. What is interesting about the field of data science is that each project can be challenging in its unique way, so you need to learn to be agile and refine the skill to learn new libraries and tools quickly depending on the project.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas  dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

In [1]:
import numpy as np 
import pandas as pd 
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests 
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
import matplotlib.cm as cm
import matplotlib.colors as colors
# import k-means from clustering stage
from sklearn.cluster import KMeans
import folium # map rendering library
from bs4 import BeautifulSoup as bs
import lxml
print('Libraries imported.')

Libraries imported.


## PART 1 : Scraping Toronto Neighbourhood from Wikipedia

For Toronto data neighbourhood data, i scrape Wikipedia file https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [2]:
r = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')
soup = bs(r.text, 'html.parser')
table=soup.find('table')
len(table)

2

In [3]:
data = table.find_all('td')
len(data)

180

In [4]:
neighbourhood = [] #init list
for i,item in enumerate(data) :
    neigh = [] #init list
    span_ = item.span.text.strip()
    if not(span_ == "Not assigned"):
        text_ = item.p.text.strip()
        poscode = text_[0:3]
        try :
            borough = item.find('a').text.strip()
        except :
            borough = "NA"
        try :
            neigh = text_.split('(')[1] #remove '('
            neigh = neigh[:-1]  #remove ')' 
            #split neighbourhood by '/' and convert to list
            #neigh = neigh.split('/')
            neigh = (((neigh.replace(' /',',')).replace(')',' ')).strip(' '))
        except:
            neigh = "NA"
        #print(f"<{poscode}><{borough}><{neigh}>")
        neigh2 = {
            "PostalCode" : poscode,
            "Borough" : borough,
            "Neighborhood" :neigh,
            }
        neighbourhood.append(neigh2)
        #end if
    #end for i
df_neigh = pd.DataFrame(neighbourhood)


In [5]:
df_neigh['Borough']=df_neigh['Borough'].replace({'Business reply mail':'East Toronto',
                                                 'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                                 'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                                 'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                                 'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})

In [6]:
df_neigh.head(5)

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [7]:
df_neigh.tail(5)

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...

In [8]:
df_neigh.shape

(103, 3)

#### we have 103 rows of Toronto Neighbourhood from the WIkipedia datasheet

## PART 2 : 

1. Reading Lat Long file from Geospatial datasheet at
https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv

In [9]:
df_latlong = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")

In [10]:
df_latlong.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [11]:
df_latlong.rename(columns={'Postal Code':'PostalCode'}, inplace=True)

In [12]:
df_neigh.head(5)

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [13]:
# merge Lat Long and create Toronto Neighbourhood dataframe
Toronto_data = df_neigh.join(df_latlong.set_index('PostalCode'), on='PostalCode')

In [14]:
Toronto_data.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  
0 -79.329656  
1 -79.315572  
2 -79.360636  
3 -79.464763  
4 -79.389494

In [15]:
Toronto_data.tail()

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999

#### new Toronto_data having information of neighborhoood and its latitude and longitude above

### PART 3:
#### Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. 


1. to add enough Markdown cells to explain what you decided to do and to report any observations you make. 
2. to generate maps to visualize your neighborhoods and how they cluster together. 

Get the coordinate for Toronto for opening map later

In [16]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### 1. Visualize Toronto Neighborhood

In [17]:
#Toronto_data = df_Toronto
# create map of Toronto using latitude and longitude values
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(Toronto_data['Latitude'], Toronto_data['Longitude'], Toronto_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_Toronto)  
    
map_Toronto

### Work with Foursquare to explore the Toronto area
### Define Foursquare Credentials and Version

Randomnly pick an area in TORONTO

In [19]:
area_no = 74
neighborhood_name = Toronto_data.loc[area_no,'Neighborhood']
neighborhood_latitude = Toronto_data.loc[area_no,'Latitude']
neighborhood_longitude = Toronto_data.loc[area_no,'Longitude']
print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Annex, North Midtown, Yorkville are 43.6727097, -79.4056784.


In [20]:
# type your answer here
radius = 500
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, neighborhood_latitude, neighborhood_longitude, VERSION, radius, LIMIT)

In [21]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '60ff83d41d562a191a280352'},
 'response': {'headerLocation': 'The Annex',
  'headerFullLocation': 'The Annex, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 21,
  'suggestedBounds': {'ne': {'lat': 43.6772097045, 'lng': -79.39946849197052},
   'sw': {'lat': 43.668209695499996, 'lng': -79.41188830802948}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ae7bc29f964a52084ad21e3',
       'name': "Ezra's Pound",
       'location': {'address': '238 Dupont St',
        'crossStreet': 'Spadina Ave',
        'lat': 43.67515283323029,
        'lng': -79.40585846415303,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.67515283323029,
          'lng': -79.40585846415303}],
        'distance': 272,
  

In [22]:
'There are {} popular spots around {}.'.format(len(results['response']['groups'][0]['items']),neighborhood_name)

'There are 21 popular spots around The Annex, North Midtown, Yorkville.'

### 2. Function to get popular venue in the neighbourhood

In [23]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
    #will put all venue in venue_list into dataframe
    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

create Toronto venus dataframe

In [24]:
Toronto_venues = getNearbyVenues(names=Toronto_data['Neighborhood'],
                                   latitudes=Toronto_data['Latitude'],
                                   longitudes=Toronto_data['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills Nort
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills Sout
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth W

In [25]:

Toronto_venues.tail()

Neighborhood  \
2148  Mimico NW, The Queensway West, South of Bloor,...   
2149  Mimico NW, The Queensway West, South of Bloor,...   
2150  Mimico NW, The Queensway West, South of Bloor,...   
2151  Mimico NW, The Queensway West, South of Bloor,...   
2152  Mimico NW, The Queensway West, South of Bloor,...   

      Neighborhood Latitude  Neighborhood Longitude  \
2148              43.628841              -79.520999   
2149              43.628841              -79.520999   
2150              43.628841              -79.520999   
2151              43.628841              -79.520999   
2152              43.628841              -79.520999   

                                      Venue  Venue Latitude  Venue Longitude  \
2148                Jim & Maria's No Frills       43.631152       -79.518617   
2149            Islington Florist & Nursery       43.630156       -79.518718   
2150  Koala Tan Tanning Salon & Sunless Spa       43.631370       -79.519006   
2151                          Value Village       43.631269       -79.518238   
2152                   Kingsway Boxing Club       43.627254       -79.526684   

              Venue Category  
2148           Grocery Store  
2149             Flower Shop  
2150           Tanning Salon  
2151  Thrift / Vintage Store  
2152                     Gym

Let's check how many venues were returned for each neighborhood

In [26]:
Toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Wilson Heights, Downsview North                        24   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      25   
Berczy Park                                                            56   
Birch Cliff, Cliffside West                                             5   
Brockton, Parkdale Village, Exhibition Place                           24   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Caledonia-Fairbanks                                                     3   
Cedarbrae                                                               8   
Central Bay Street                                                     70   
Christie                                                               16   
Church and Wellesley                                                   78   
Clairville, Humberwood, Woodbine Downs, West Hu...                      5   
Clarks Corners, Tam O'Shanter, Sullivan                                13   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                       10   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills Nort                                                          6   
Don Mills Sout                                                         20   
Dorset Park, Wexford Heights, Scarborough Town ...                      6   
Downsview Centra                                                        3   
Downsview East                                                          3   
Downsview Northwes                                                      4   
Downsview Wes                                                           4   
Dufferin, Dovercourt Village                                           16   
Enclave of L4W                                                         14   
Enclave of M4L                                                         16   
Enclave of M5E                                                         98   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                        10   
Guildwood, Morningside, West Hill                                       8   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          24   
Hillcrest Village                                                       5   
Humber Summit                                                           3   
Humberlea, Emery                                                        2   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         23   
Kennedy Park, Ionview, East Birchmount Park                             5   
Kensington Market, Chinatown, Grange Park                              67   
Kingsview Village, St. Phillips, Martin Grove G...                      4   
Lawrence Manor, Lawrence Heights                

Let's find out how many unique categories can be curated from all the returned venues

In [27]:
print('There are {} uniques categories.'.format(len(Toronto_venues['Venue Category'].unique())))

There are 277 uniques categories.


### 3. Analyze Each Neighborhood

In [28]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']])

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighborhood  Venue Category_Accessories Store  \
0         Parkwoods                                 0   
1         Parkwoods                                 0   
2         Parkwoods                                 0   
3  Victoria Village                                 0   
4  Victoria Village                                 0   

   Venue Category_Adult Boutique  Venue Category_Airport  \
0                              0                       0   
1                              0                       0   
2                              0                       0   
3                              0                       0   
4                              0                       0   

   Venue Category_Airport Food Court  Venue Category_Airport Gate  \
0                                  0                            0   
1                                  0                            0   
2                                  0                            0   
3                                  0                            0   
4                                  0                            0   

   Venue Category_Airport Lounge  Venue Category_Airport Service  \
0                              0                               0   
1                              0                               0   
2                              0                               0   
3                              0                               0   
4                              0                               0   

   Venue Category_Airport Terminal  Venue Category_American Restaurant  \
0                                0                                   0   
1                                0                                   0   
2                                0                                   0   
3                                0                                   0   
4                                0                                   0   

   Venue Category_Antique Shop  Venue Category_Aquarium  \
0                            0                        0   
1                            0                        0   
2                            0                        0   
3                            0                        0   
4                            0                        0   

   Venue Category_Art Gallery  Venue Category_Art Museum  \
0                           0                          0   
1                           0                          0   
2                           0                          0   
3                           0                          0   
4                           0                          0   

   Venue Category_Arts & Crafts Store  Venue Category_Asian Restaurant  \
0                                   0                                0   
1                                   0                                0   
2                                   0                                0   
3                                   0                                0   
4                                   0                                0   

   Venue Category_Athletics & Sports  Venue Category_Auto Garage  \
0                                  0                           0   
1                                  0                           0   
2                                  0                           0   
3                                  0                           0   
4                                  0                           0   

   Venue Category_Auto Workshop  Venue Category_BBQ Joint  \
0                             0                         0   
1                             0                         0   
2                             0                         0   
3                             0                         0   
4                             0                         0   

   Venue Category_Baby Store  Venue Category_Bagel Shop  \
0                          0          

And let's examine the new dataframe size.

In [29]:
Toronto_onehot.shape

(2153, 278)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [30]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  \
0                                           Agincourt   
1                              Alderwood, Long Branch   
2     Bathurst Manor, Wilson Heights, Downsview North   
3                                     Bayview Village   
4                   Bedford Park, Lawrence Manor East   
5                                         Berczy Park   
6                         Birch Cliff, Cliffside West   
7        Brockton, Parkdale Village, Exhibition Place   
8   CN Tower, King and Spadina, Railway Lands, Har...   
9                                 Caledonia-Fairbanks   
10                                          Cedarbrae   
11                                 Central Bay Street   
12                                           Christie   
13                               Church and Wellesley   
14  Clairville, Humberwood, Woodbine Downs, West H...   
15            Clarks Corners, Tam O'Shanter, Sullivan   
16    Cliffside, Cliffcrest, Scarborough Village West   
17                     Commerce Court, Victoria Hotel   
18                                         Davisville   
19                                   Davisville North   
20   Del Ray, Mount Dennis, Keelsdale and Silverthorn   
21                                     Don Mills Nort   
22                                     Don Mills Sout   
23  Dorset Park, Wexford Heights, Scarborough Town...   
24                                   Downsview Centra   
25                                     Downsview East   
26                                 Downsview Northwes   
27                                      Downsview Wes   
28                       Dufferin, Dovercourt Village   
29                                     Enclave of L4W   
30                                     Enclave of M4L   
31                                     Enclave of M5E   
32  Eringate, Bloordale Gardens, Old Burnhamthorpe...   
33                       Fairview, Henry Farm, Oriole   
34             First Canadian Place, Underground city   
35                           Forest Hill North & West   
36                           Garden District, Ryerson   
37                                          Glencairn   
38                    Golden Mile, Clairlea, Oakridge   
39                  Guildwood, Morningside, West Hill   
40  Harbourfront East, Union Station, Toronto Islands   
41                      High Park, The Junction South   
42                                  Hillcrest Village   
43                                      Humber Summit   
44                                   Humberlea, Emery   
45                                 Humewood-Cedarvale   
46                     India Bazaar, The Beaches West   
47        Kennedy Park, Ionview, East Birchmount Park   
48          Kensington Market, Chinatown, Grange Park   
49  Kingsview Village, St. Phillips, Martin Grove ...   
50                   Lawrence Manor, Lawrence Heights   
51                                      Lawrence Park   
52                                            Leaside   
53                           Little Portugal, Trinity   
54                                     Malvern, Rouge   
55  Milliken, Agincourt North, Steeles East, L'Amo...   
56  Mimico NW, The Queensway West, South of Bloor,...   
57                        Moore Park, Summerhill East   
58       New Toronto, Mimico South, Humber Bay Shores   
59           North Park, Maple Leaf Park, Upwood Park   
60                                 North Toronto West   
61                    Northwood Park, York University   
62  Old Mill South, King's Mill Park, Sunnylea, Hu...   
63                      Ontario Provincial Government   
64                             Parkdale, Roncesvalles   
65                    Parkview Hill, Woodbine Gardens   
66                                          Parkwoods   
67                          Regent Park, Harbourfront   
68                           Richmond, Adelaide, King   
69                                           Roseda

In [31]:
Toronto_grouped.shape

(100, 278)

#### Let's print each neighborhood along with the top 5 most common venues

In [32]:
num_top_venues = 5

for hood in Toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = Toronto_grouped[Toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                                      venue  freq
0  Venue Category_Latin American Restaurant  0.25
1               Venue Category_Skating Rink  0.25
2             Venue Category_Breakfast Spot  0.25
3                     Venue Category_Lounge  0.25
4        Venue Category_Moroccan Restaurant  0.00


----Alderwood, Long Branch----
                           venue  freq
0     Venue Category_Pizza Place  0.25
1     Venue Category_Coffee Shop  0.12
2    Venue Category_Skating Rink  0.12
3  Venue Category_Sandwich Place  0.12
4             Venue Category_Pub  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
                                      venue  freq
0                       Venue Category_Bank  0.08
1                Venue Category_Coffee Shop  0.08
2                Venue Category_Gas Station  0.04
3                  Venue Category_Gift Shop  0.04
4  Venue Category_Middle Eastern Restaurant  0.04


----Bayview Village----
                               

4  Venue Category_Accessories Store  0.00


----Garden District, Ryerson----
                            venue  freq
0      Venue Category_Coffee Shop  0.11
1   Venue Category_Clothing Store  0.09
2            Venue Category_Hotel  0.03
3  Venue Category_Bubble Tea Shop  0.03
4   Venue Category_Cosmetics Shop  0.03


----Glencairn----
                                venue  freq
0  Venue Category_Japanese Restaurant   0.2
1               Venue Category_Bakery   0.2
2        Venue Category_Metro Station   0.2
3     Venue Category_Asian Restaurant   0.2
4                 Venue Category_Park   0.2


----Golden Mile, Clairlea, Oakridge----
                         venue  freq
0      Venue Category_Bus Line   0.2
1        Venue Category_Bakery   0.2
2  Venue Category_Soccer Field   0.1
3          Venue Category_Park   0.1
4  Venue Category_Intersection   0.1


----Guildwood, Morningside, West Hill----
                               venue  freq
0      Venue Category_Medical Center  0.12
1    

                              venue  freq
0               Venue Category_Café  0.14
1     Venue Category_Sandwich Place  0.14
2        Venue Category_Coffee Shop  0.14
3           Venue Category_Pharmacy  0.05
4  Venue Category_Indian Restaurant  0.05


----The Beaches----
                                  venue  freq
0      Venue Category_Health Food Store  0.25
1                    Venue Category_Pub  0.25
2                  Venue Category_Trail  0.25
3           Venue Category_Neighborhood  0.25
4  Venue Category_Performing Arts Venue  0.00


----The Danforth  East----
                              venue  freq
0  Venue Category_Convenience Store  0.33
1      Venue Category_Metro Station  0.33
2               Venue Category_Park  0.33
3  Venue Category_Accessories Store  0.00
4  Venue Category_Mobile Phone Shop  0.00


----The Danforth West, Riverdale----
                                   venue  freq
0        Venue Category_Greek Restaurant  0.17
1      Venue Category_Italian Restau

Let's put that into a pandas dataframe

In [33]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [34]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  \
0                                        Agincourt   
1                           Alderwood, Long Branch   
2  Bathurst Manor, Wilson Heights, Downsview North   
3                                  Bayview Village   
4                Bedford Park, Lawrence Manor East   

                      1st Most Common Venue        2nd Most Common Venue  \
0  Venue Category_Latin American Restaurant  Venue Category_Skating Rink   
1                Venue Category_Pizza Place   Venue Category_Coffee Shop   
2                       Venue Category_Bank   Venue Category_Coffee Shop   
3        Venue Category_Japanese Restaurant          Venue Category_Café   
4         Venue Category_Italian Restaurant   Venue Category_Coffee Shop   

           3rd Most Common Venue                   4th Most Common Venue  \
0  Venue Category_Breakfast Spot                   Venue Category_Lounge   
1    Venue Category_Skating Rink           Venue Category_Sandwich Place   
2     Venue Category_Gas Station                Venue Category_Gift Shop   
3            Venue Category_Bank       Venue Category_Chinese Restaurant   
4  Venue Category_Sandwich Place  Venue Category_Comfort Food Restaurant   

                      5th Most Common Venue  \
0        Venue Category_Moroccan Restaurant   
1                        Venue Category_Pub   
2  Venue Category_Middle Eastern Restaurant   
3          Venue Category_Accessories Store   
4                Venue Category_Pizza Place   

                6th Most Common Venue  \
0  Venue Category_Monument / Landmark   
1             Venue Category_Pharmacy   
2        Venue Category_Shopping Mall   
3        Venue Category_Movie Theater   
4         Venue Category_Cupcake Shop   

                            7th Most Common Venue  \
0  Venue Category_Molecular Gastronomy Restaurant   
1                              Venue Category_Gym   
2                Venue Category_Mobile Phone Shop   
3                            Venue Category_Motel   
4                              Venue Category_Pub   

                       8th Most Common Venue  \
0  Venue Category_Modern European Restaurant   
1      Venue Category_Peking Duck Restaurant   
2              Venue Category_Sandwich Place   
3         Venue Category_Moroccan Restaurant   
4                  Venue Category_Restaurant   

                9th Most Common Venue  \
0    Venue Category_Mobile Phone Shop   
1   Venue Category_Persian Restaurant   
2  Venue Category_Fried Chicken Joint   
3  Venue Category_Monument / Landmark   
4                 Venue Category_Café   

                           10th Most Common Venue  
0               Venue Category_Miscellaneous Shop  
1         Venue Category_Mediterranean Restaurant  
2               Venue Category_Frozen Yogurt Shop  
3  Venue Category_Molecular Gastronomy Restaurant  
4                          Venue Category_Butcher

### 4. Cluster Neighborhoods
#### Run k-means to cluster the neighborhood into 5 clusters.

In [36]:
neighborhoods_venues_sorted['1st Most Common Venue'].unique()

array(['Venue Category_Latin American Restaurant',
       'Venue Category_Pizza Place', 'Venue Category_Bank',
       'Venue Category_Japanese Restaurant',
       'Venue Category_Italian Restaurant', 'Venue Category_Bakery',
       'Venue Category_College Stadium', 'Venue Category_Café',
       'Venue Category_Airport Terminal', "Venue Category_Women's Store",
       'Venue Category_Thai Restaurant', 'Venue Category_Coffee Shop',
       'Venue Category_Grocery Store', 'Venue Category_Bar',
       'Venue Category_Motel', 'Venue Category_Sandwich Place',
       'Venue Category_Hotel', 'Venue Category_Discount Store',
       'Venue Category_Caribbean Restaurant',
       'Venue Category_Clothing Store',
       'Venue Category_Indian Restaurant', 'Venue Category_Food Truck',
       'Venue Category_Airport', 'Venue Category_Athletics & Sports',
       'Venue Category_Pharmacy', 'Venue Category_Light Rail Station',
       'Venue Category_Jewelry Store', 'Venue Category_Bus Line',
       'Venu

In [37]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [38]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = Toronto_data

# merge Toronto_grouped with Toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.tail() # check the last columns!

PostalCode           Borough  \
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  \
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944   
99                                Church and Wellesley  43.665860 -79.383160   
100                                     Enclave of M4L  43.662744 -79.321558   
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509   
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999   

     Cluster Labels              1st Most Common Venue  \
98              0.0          Venue Category_Smoke Shop   
99              0.0         Venue Category_Coffee Shop   
100             0.0  Venue Category_Light Rail Station   
101             2.0      Venue Category_Baseball Field   
102             0.0      Venue Category_Hardware Store   

                     2nd Most Common Venue  \
98                     Venue Category_Pool   
99         Venue Category_Sushi Restaurant   
100             Venue Category_Pizza Place   
101       Venue Category_Accessories Store   
102  Venue Category_Thrift / Vintage Store   

                   3rd Most Common Venue       4th Most Common Venue  \
98                   Venue Category_Park        Venue Category_River   
99    Venue Category_Japanese Restaurant      Venue Category_Gay Bar   
100         Venue Category_Burrito Place      Venue Category_Brewery   
101     Venue Category_Mobile Phone Shop        Venue Category_Motel   
102  Venue Category_Fast Food Restaurant  Venue Category_Wings Joint   

                  5th Most Common Venue                6th Most Common Venue  \
98    Venue Category_Miscellaneous Shop   Venue Category_Moroccan Restaurant   
99            Venue Category_Restaurant  Venue Category_Fast Food Restaurant   
100        Venue Category_Auto Workshop            Venue Category_Restaurant   
101  Venue Category_Moroccan Restaurant   Venue Category_Monument / Landmark   
102          Venue Category_Flower Shop     Venue Category_Convenience Store   

                              7th Most Common Venue  \
98               Venue Category_Monument / Landmark   
99                               Venue Category_Pub   
100                    Venue Category_Garden Center   
101  Venue Category_Molecular Gastronomy Restaurant   
102                     Venue Category_Burger Joint   

                              8th Most Common Venue  \
98   Venue Category_Molecular Gastronomy Restaurant   
99                       Venue Category_Men's Store   
100                       Venue Category_Skate Park   
101       Venue Category_Modern European Restaurant   
102                    Venue Category_Tanning Salon   

                         9th Most Common Venue  \
98   Venue Category_Modern European Restaurant   
99     Venue Category_Mediterranean Restaurant   
100                      Venue Category_Garden   
101          Venue Category_Miscellaneous Shop   
102                      Venue Category_Bakery   

                  10th Most Common Venue  
98      Venue Category_Mobile Phone Shop  
99            Venue Category_Yoga Studio  
100  Venue Category_Gym / Fitness Center  
101                Venue Category_Museum  
102                   Venue Category_Gym

Finally, let's visualize the resulting clusters

In [39]:
Toronto_merged['Cluster Labels'].unique()

array([ 3.,  0., nan,  4.,  1.,  2.])

#simply drop whole row with NaN in "Most Common Venue" column

In [40]:
# simply drop whole row with NaN in "Most Common Venue" column
Toronto_merged.dropna(subset=["1st Most Common Venue"], axis=0, inplace=True)

# reset index, because we droped two rows
Toronto_merged.reset_index(drop=True, inplace=True)

In [41]:
Toronto_merged['Cluster Labels'].unique()

array([3., 0., 4., 1., 2.])

In [43]:
#change the Cluster Labels to int type for map visualization
Toronto_merged['Cluster Labels'] = Toronto_merged['Cluster Labels'].astype(int)

In [44]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighborhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### Cluster 1

In [45]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels  \
1         North York               0   
2   Downtown Toronto               0   
3         North York               0   
4       Queen's Park               0   
5        Scarborough               0   
6         North York               0   
7          East York               0   
8   Downtown Toronto               0   
9         North York               0   
10         Etobicoke               0   
12        North York               0   
13         East York               0   
14  Downtown Toronto               0   
15              York               0   
16         Etobicoke               0   
17       Scarborough               0   
18       The Beaches               0   
19  Downtown Toronto               0   
21       Scarborough               0   
22         East York               0   
23  Downtown Toronto               0   
24  Downtown Toronto               0   
25       Scarborough               0   
26        North York               0   
27        North York               0   
28         East York               0   
29  Downtown Toronto               0   
30          Dufferin               0   
32        North York               0   
33        North York               0   
35  Downtown Toronto               0   
36   Little Portugal               0   
37       Scarborough               0   
38        North York               0   
40      The Danforth               0   
41  Downtown Toronto               0   
42          Brockton               0   
43       Scarborough               0   
44        North York               0   
45      India Bazaar               0   
46  Downtown Toronto               0   
47        North York               0   
48        North York               0   
49       Scarborough               0   
51        North York               0   
52                NA               0   
53        North York               0   
54              York               0   
56       Scarborough               0   
57        North York               0   
58        North York               0   
60                NA               0   
61              York               0   
63       Scarborough               0   
65        Davisville               0   
66       Forest Hill               0   
67         High Park               0   
68         Etobicoke               0   
69       Scarborough               0   
70        North York               0   
71     North Toronto               0   
72         The Annex               0   
73          Parkdale               0   
74       Mississauga               0   
75         Etobicoke               0   
76       Scarborough               0   
77        Davisville               0   
78  Downtown Toronto               0   
79         Runnymede               0   
80       Scarborough               0   
82  Downtown Toronto               0   
83       Scarborough               0   
84        Summerhill               0   
85  Downtown Toronto               0   
86         Etobicoke               0   
87         Etobicoke               0   
88       Scarborough               0   
90  Downtown Toronto               0   
91         Etobicoke               0   
92         Etobicoke               0   
93  Downtown Toronto               0   
94  Downtown Toronto               0   
95         Etobicoke               0   
96  Downtown Toronto               0   
97      East Toronto               0   
99         Etobicoke               0   

                        1st Most Common Venue  \
1                  Venue Category_Pizza Place   
2                  Venue Category_Coffee Shop   
3               Venue Category_Clothing Store   
4                  Venue Category_Coffee Shop   
5         Venue Category_Fast Food Restaurant   
6         Venue Category_Caribbean Restaurant   
7                  Venue Category_Pizza Place   
8                  Venue Category_Coffee Shop   
9          Venue Category_Japanese Restaurant   
10         Venue Category_Filipino Restaur

#### Cluster 2

In [46]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
31  Scarborough               1  Venue Category_Playground   

               2nd Most Common Venue         3rd Most Common Venue  \
31  Venue Category_Mobile Phone Shop  Venue Category_Movie Theater   

   4th Most Common Venue               5th Most Common Venue  \
31  Venue Category_Motel  Venue Category_Moroccan Restaurant   

                 6th Most Common Venue  \
31  Venue Category_Monument / Landmark   

                             7th Most Common Venue  \
31  Venue Category_Molecular Gastronomy Restaurant   

                        8th Most Common Venue  \
31  Venue Category_Modern European Restaurant   

                9th Most Common Venue              10th Most Common Venue  
31  Venue Category_Miscellaneous Shop  Venue Category_Martial Arts School

#### Cluster 3

In [47]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels                  1st Most Common Venue  \
55  North York               2  Venue Category_Furniture / Home Store   
98   Etobicoke               2          Venue Category_Baseball Field   

               2nd Most Common Venue             3rd Most Common Venue  \
55     Venue Category_Baseball Field  Venue Category_Accessories Store   
98  Venue Category_Accessories Store  Venue Category_Mobile Phone Shop   

                 4th Most Common Venue               5th Most Common Venue  \
55  Venue Category_Martial Arts School                Venue Category_Motel   
98                Venue Category_Motel  Venue Category_Moroccan Restaurant   

                 6th Most Common Venue  \
55  Venue Category_Moroccan Restaurant   
98  Venue Category_Monument / Landmark   

                             7th Most Common Venue  \
55              Venue Category_Monument / Landmark   
98  Venue Category_Molecular Gastronomy Restaurant   

                             8th Most Common Venue  \
55  Venue Category_Molecular Gastronomy Restaurant   
98       Venue Category_Modern European Restaurant   

                        9th Most Common Venue  \
55  Venue Category_Modern European Restaurant   
98          Venue Category_Miscellaneous Shop   

              10th Most Common Venue  
55  Venue Category_Mobile Phone Shop  
98             Venue Category_Museum

#### Cluster 4

In [48]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels                1st Most Common Venue  \
0         North York               3  Venue Category_Fast Food Restaurant   
20              York               3         Venue Category_Women's Store   
34         East York               3     Venue Category_Convenience Store   
39        North York               3               Venue Category_Airport   
50        North York               3                  Venue Category_Park   
59     Lawrence Park               3              Venue Category_Bus Line   
62              York               3     Venue Category_Convenience Store   
64        North York               3     Venue Category_Convenience Store   
81        Moore Park               3                  Venue Category_Park   
89  Downtown Toronto               3                  Venue Category_Park   

               2nd Most Common Venue              3rd Most Common Venue  \
0                Venue Category_Park   Venue Category_Food & Drink Shop   
20               Venue Category_Pool                Venue Category_Park   
34      Venue Category_Metro Station                Venue Category_Park   
39   Venue Category_Business Service                Venue Category_Park   
50  Venue Category_Accessories Store  Venue Category_Miscellaneous Shop   
59               Venue Category_Park         Venue Category_Swim School   
62               Venue Category_Park   Venue Category_Accessories Store   
64               Venue Category_Park   Venue Category_Accessories Store   
81        Venue Category_Summer Camp   Venue Category_Accessories Store   
89         Venue Category_Playground               Venue Category_Trail   

                       4th Most Common Venue  \
0   Venue Category_Middle Eastern Restaurant   
20          Venue Category_Accessories Store   
34          Venue Category_Accessories Store   
39          Venue Category_Accessories Store   
50                      Venue Category_Motel   
59          Venue Category_Accessories Store   
62         Venue Category_Miscellaneous Shop   
64         Venue Category_Miscellaneous Shop   
81  Venue Category_Middle Eastern Restaurant   
89          Venue Category_Accessories Store   

                 5th Most Common Venue               6th Most Common Venue  \
0   Venue Category_Moroccan Restaurant  Venue Category_Monument / Landmark   
20   Venue Category_Miscellaneous Shop                Venue Category_Motel   
34    Venue Category_Mobile Phone Shop                Venue Category_Motel   
39    Venue Category_Mobile Phone Shop                Venue Category_Motel   
50  Venue Category_Moroccan Restaurant  Venue Category_Monument / Landmark   
59    Venue Category_Mobile Phone Shop  Venue Category_Moroccan Restaurant   
62                Venue Category_Motel  Venue Category_Moroccan Restaurant   
64                Venue Category_Motel  Venue Category_Moroccan Restaurant   
81  Venue Category_Moroccan Restaurant  Venue Category_Monument / Landmark   
89    Venue Category_Mobile Phone Shop                Venue Category_Motel   

                             7th Most Common Venue  \
0   Venue Category_Molecular Gastronomy Restaurant   
20              Venue Category_Moroccan Restaurant   
34              Venue Category_Moroccan Restaurant   
39              Venue Category_Moroccan Restaurant   
50  Venue Category_Molecular Gastronomy Restaurant   
59              Venue Category_Monument / Landmark   
62              Venue Category_Monument / Landmark   
64              Venue Category_Monument / Landmark   
81  Venue Category_Molecular Gastronomy Restaurant   
89              Venue Category_Moroccan Restaurant   

                             8th Most Common Venue  \
0        Venue Category_Modern European Restaurant   
20              Venue Category_Monument / Landmark   
34              Venue Category_Monument / Landmark   
39              Venue Category_Monument / Landmark   
50       Venue Category_Modern European Restaurant   
59  Venue Category_Molecular Gas

#### Cluster 5

In [49]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
11  Scarborough               4    Venue Category_Bar   

               2nd Most Common Venue             3rd Most Common Venue  \
11  Venue Category_Accessories Store  Venue Category_Mobile Phone Shop   

   4th Most Common Venue               5th Most Common Venue  \
11  Venue Category_Motel  Venue Category_Moroccan Restaurant   

                 6th Most Common Venue  \
11  Venue Category_Monument / Landmark   

                             7th Most Common Venue  \
11  Venue Category_Molecular Gastronomy Restaurant   

                        8th Most Common Venue  \
11  Venue Category_Modern European Restaurant   

                9th Most Common Venue 10th Most Common Venue  
11  Venue Category_Miscellaneous Shop  Venue Category_Museum

### Cluster 1 and Cluster 4 are having most popular venues, while Cluster 2 , Cluster 3 and Cluster 5 being less popular